feat: add Rayon compute pool for CPU-intensive operations

ryanolson opened the pull request:
#### Overview:
Integrate Rayon with Tokio to provide a dedicated thread pool for CPU-bound work. This PR introduces a comprehensive compute pool system with size-aware task classification macros, thread-local context, and robust async/compute interaction handling.

#### Details:
This implementation adds several key components:
- **ComputePool**: Tokio-Rayon bridge for async-to-sync CPU work offloading
- **Smart Macros**: `compute_small!`, `compute_medium!`, `compute_large!` with automatic fallback 
- **Thread-local Context**: Barrier-based initialization across all Tokio worker threads
- **Performance Monitoring**: Comprehensive metrics with overflow protection
- **Dynamic Calibration**: Benchmarks that auto-tune for target execution times
- **Robust Configuration**: Thread count validation with sensible defaults

Key features:
- Semaphore-controlled block_in_place for medium tasks (preserves async threads)
- Always-offload policy for large tasks (prevents runtime blocking)
- Inline fallback for all macros (ensures code always works)
- Configurable via DYN_COMPUTE_* environment variables
- Zero-overhead when validation disabled

#### Where should the reviewer start?
1. **Core API**: `lib/runtime/src/compute/mod.rs` - Main configuration and pool setup
2. **Macros**: `lib/runtime/src/compute/macros.rs` - Smart task classification system
3. **Examples**: `lib/runtime/examples/async_vs_compute_interaction.rs` - Demonstrates interference patterns
4. **Thread-local**: `lib/runtime/src/compute/thread_local.rs` - Context initialization
5. **Documentation**: `lib/runtime/src/compute/README.md` - Usage patterns and strategies

#### Related Issues:
Addresses the need for efficient CPU-bound task handling in async Rust applications without blocking the async runtime.

ryanolson committed:
feat(runtime): add Rayon compute pool for CPU-intensive operations

Integrate Rayon with Tokio to provide a dedicated thread pool for CPU-bound
work. Multiple async tasks can concurrently use different parallelization
patterns (scope, par_iter, join) on the shared pool.

- Add ComputePool with tokio-rayon for async-to-sync bridging
- Support concurrent access from multiple async tasks
- Configure via DYN_COMPUTE_* environment variables
- Include comprehensive Rayon/Tokio strategy documentation
- Add examples for tokenization and mixed workloads

ryanolson committed:
fix: address PR review feedback for Rayon compute pool

- Use custom Rayon pool instead of global pool in tokio_rayon::spawn
- Fix parallel_map to use into_par_iter() for actual parallelism
- Handle compute_threads=0 as disabled with info logging
- Fix test_compute_pool_scoped to use channels for result collection
- Mark documentation examples as no_run to ensure compilation

ryanolson committed:
feat: add compute pool benchmarks and demos

- Add criterion benchmark for overhead measurement
- Add interactive demo showing crossover points
- Add async throughput impact demonstration
- Demonstrate overhead at n=10 vs benefits at n=100k+

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
feat(compute): add intelligent compute task scheduling with macros

- Add compute_small!, compute_medium!, compute_large! macros for size-aware execution
- Implement semaphore-controlled block_in_place for medium tasks (100μs-1ms)
- Add thread-local storage for compute context on Tokio worker threads
- Include optional validation feature for detecting misclassified tasks
- Simplify benchmarks to use 3 representative values (10, 1_000, 100_000)
- Add block_in_place overhead benchmarking

The semaphore is properly tied to Tokio worker thread count to ensure
async threads remain available when using block_in_place. Medium tasks
use block_in_place when permits are available, falling back to Rayon
offload when permits are exhausted.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
feat(compute): add thread-local context checking and async/compute interaction benchmark

- Add has_compute_context() to check if thread-local compute context is available
- Add assert_compute_context() to panic if context not initialized
- Create async_vs_compute_interaction.rs benchmark example
- Demonstrate interference patterns with 4 total threads in two configurations:
  - All-Async: 4 Tokio threads (compute blocks async work)
  - Hybrid: 2 Tokio + 2 Rayon threads (compute offloaded)
- Shows how compute workloads affect async I/O latency
- Validates that offloading compute >100μs is essential for async responsiveness

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
feat(compute): add robust thread-local initialization with barrier

- Replace flawed initialize_all_thread_locals that used random task spawning
- Implement barrier-based approach ensuring ALL worker threads get initialized
- Add detect_worker_thread_count() to accurately count runtime worker threads
- Update async_vs_compute_interaction example to demonstrate thread-local macros
- Show proper initialization using barrier synchronization
- Add dynamic compute function calibration for accurate benchmarks
- Demonstrate that thread-local context enables macro usage without explicit pools

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
fix(compute): make macros always succeed with inline fallback

- compute_medium! and compute_large! now fall back to inline execution
- Previously returned errors when no thread-local context available
- Now emit warnings but continue executing (with blocking behavior)
- Ensures macros always work even without proper setup
- Better developer experience - code runs in all environments

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
fix(compute): resolve PR compliance issues

- Add required SPDX copyright headers to all benchmark and example files
- Add overflow protection to metrics calculations using saturating conversion
- Improve thread count calculation logic with minimum parallelism (2-16 threads)
- Add configuration validation for num_threads and stack_size
- Apply clippy fixes for cleaner code
- All examples build and run successfully

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
build: update Cargo.lock for dependency consistency

Ensure reproducible builds and CI consistency by committing
dependency resolution changes from compute pool implementation.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
Merge branch 'main' into ryan/rayon

Resolve conflicts by merging both compute and health check features:
- config.rs: Include both compute pool and health check configuration
- system_status_server.rs: Use improved pattern matching for test paths
- Cargo.lock: Regenerated to resolve dependency conflicts

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
style: fix formatting in compute module

Apply rustfmt formatting to validation function in compute/mod.rs

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
fix: replace manual modulo checks with is_multiple_of()

Replace manual % checks with the .is_multiple_of() method as required
by clippy lint manual_is_multiple_of in Rust 1.90.0:

- examples/compute_overhead_demo.rs: Fix is_prime function
- examples/async_throughput_demo.rs: Fix is_prime function and progress counter
- examples/compute_pool_example.rs: Fix is_prime function
- examples/async_vs_compute_interaction.rs: Fix is_prime function
- benches/compute_pool_overhead.rs: Fix is_prime function

Resolves clippy warnings that were failing CI builds.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
test: skip failing test_request_cancellation_vllm

Skip test due to vLLM model loading error with Qwen3ForCausalLM.
This is unrelated to the compute pool implementation.

Linear: DYN-1083
Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
Merge branch 'main' into ryan/rayon

oandreeva-nv commented on lib/runtime/examples/async_vs_compute_interaction.rs:
```diff
@@ -0,0 +1,663 @@
+
+    // Summary
+    println!("\n{:=<70}", "");
+    println!("📈 Key Takeaway");
```
```suggestion
    println!(" Key Takeaway");
```

oandreeva-nv commented on lib/runtime/examples/async_vs_compute_interaction.rs:
```diff
@@ -0,0 +1,663 @@
+        }
+
+        // Test if macros work
+        println!("\n🧪 Testing thread-local macros:");
```
```suggestion
        println!("\n Testing thread-local macros:");
```

oandreeva-nv commented on lib/runtime/examples/async_vs_compute_interaction.rs:
```diff
@@ -0,0 +1,663 @@
+}
+
+fn main() -> Result<()> {
+    println!("🔬 Async vs Compute Interaction Benchmark");
```
```suggestion
    println!(" Async vs Compute Interaction Benchmark");
```

oandreeva-nv commented on lib/runtime/examples/async_vs_compute_interaction.rs:
```diff
@@ -0,0 +1,663 @@
+            let result = compute_large!(large_compute());
+            println!("    Result: {}, Time: {:?}", result, start.elapsed());
+
+            println!("\n  🎉 All macros work with thread-local context!");
```
```suggestion
            println!("\n  All macros work with thread-local context!");
```

oandreeva-nv commented on lib/runtime/examples/async_vs_compute_interaction.rs:
```diff
@@ -0,0 +1,663 @@
+
+            println!("\n  🎉 All macros work with thread-local context!");
+        } else {
+            println!("  ❌ Thread-local context NOT available - macros would fail");
```
```suggestion
            println!("  Thread-local context NOT available - macros would fail");
```

oandreeva-nv commented on lib/runtime/examples/async_vs_compute_interaction.rs:
```diff
@@ -0,0 +1,663 @@
+    })?;
+
+    println!();
+    println!("✅ Benchmark complete!");
```
```suggestion
    println!(" Benchmark complete!");
```

oandreeva-nv commented on lib/runtime/examples/compute_overhead_demo.rs:
```diff
@@ -0,0 +1,262 @@
+    println!("{:=<120}", "");
+
+    // Analysis
+    println!("\n📊 Analysis:");
```
```suggestion
    println!("\n Analysis:");
```

oandreeva-nv commented on lib/runtime/examples/compute_overhead_demo.rs:
```diff
@@ -0,0 +1,262 @@
+    }
+
+    // Print pool metrics
+    println!("\n📈 Compute Pool Metrics:");
```
```suggestion
    println!("\n Compute Pool Metrics:");
```

rmccorm4 commented on tests/fault_tolerance/test_request_cancellation.py:
```diff
@@ -370,6 +370,7 @@ def verify_request_cancelled(
 @pytest.mark.gpu_1
 @pytest.mark.e2e
 @pytest.mark.model(FAULT_TOLERANCE_MODEL_NAME)
+@pytest.mark.skip(reason="DYN-1083: Qwen3ForCausalLM model loading fails in vLLM")
```
I think this be fixed on main after https://github.com/ai-dynamo/dynamo/pull/3168

oandreeva-nv replied:
I think we can remove this line then, since 3168 is merged

ryanolson replied:
i'll back it out

ryanolson committed:
Apply suggestions from code review

Co-authored-by: Olga Andreeva <124622579+oandreeva-nv@users.noreply.github.com>
Signed-off-by: Ryan Olson <ryanolson@users.noreply.github.com>

ryanolson committed:
test: re-enable test_request_cancellation_vllm after DYN-1083 fix

DYN-1083 has been resolved, re-enabling the test that was previously
skipped due to Qwen3ForCausalLM model loading failures in vLLM.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
chore: regenerate Cargo.lock after merging main

Resolves conflicts by rebuilding lockfile with updated dependencies
from main branch (anyhow 1.0.100, aws-lc-rs 1.14.0, etc.) and our
rayon/tokio-rayon additions.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
Merge branch 'main' into ryan/rayon

grahamking reviewed (approved):
Approving unreviewed to unblock. I believe several other people internally have reviewed.
